source/common/network/proxy_protocol_filter_state.cc

```diff
@@ -9,5 +12,66 @@ const std::string& ProxyProtocolFilterState::key() {
   CONSTRUCT_ON_FIRST_USE(std::string, "envoy.network.proxy_protocol_options");
 }
 
+class ProxyProtocolFilterStateReflection : public StreamInfo::FilterState::ObjectReflection {
+public:
+  ProxyProtocolFilterStateReflection(const ProxyProtocolFilterState* object) : object_(object) {}
+
+  FieldType getField(absl::string_view tlv_type_str) const override {    
+    // Specified tlv_type must be parsable as an int.
+    ASSERT(!tlv_type_str.empty());
+    int tlv_type;
+    if (!absl::SimpleAtoi(tlv_type_str, &tlv_type)) {
+    throw EnvoyException(fmt::format(
+        "Invalid parameter provided for FIELD value: {}. The proxy protocol TLV type must be parsable as int.",
+        tlv_type_str));
+    }
+
+    // Check if a valid TLV type was passed in.
+    if (tlv_type >= 256 || tlv_type <= 0) {
+      throw EnvoyException(fmt::format("Invalid parameter provided for FIELD value: "
+                                      "{}. The proxy protocol TLV type must be a positive integer less than 256.",
+                                      tlv_type_str));
+    }
+
+    // Parse the TLVs with the given type from the filter state object.
+    std::ostringstream oss;
+    int match_count = 0;
+    for (auto& tlv : object_->value().tlv_vector_) {
+      if (tlv.type == tlv_type) {
+        if (match_count > 0)
+          oss << ", ";
+        oss << Base64::encode(reinterpret_cast<const char*>(tlv.value.data()), tlv.value.size());
+        match_count++;
+      }
+    }
+    return oss.str();
```

Comment:
I think you cannot return string here directly. The return type `using FieldType = absl::variant<absl::monostate, absl::string_view, int64_t>;` only accept `absl::string_view` or number.
If string is returned, a temporary copy will be refered by the string view. It's dangling reference.

Maybe you can only log first value that related to the `tlv_type` for now.

(I think we can refactor the return type to accept both string_view or string. But it need more careful design.

Reply:
See https://github.com/envoyproxy/envoy/issues/29531

Reply:
Related: why do you need to base 64 encode it? The recipient should handle it as byte array.

Reply:
@kyessenov Base64 encoding the TLV value makes it possible to safely pass the binary data as a text HTTP header. I wasn't aware that it was possible to pass anything other than text in an HTTP header. Would you mind clarifying what you're envisioning, and why it's a preferable approach?

Reply:
I see, I wonder if it's better done at the consumption place: there's `raw_value` field in HeaderOperations but it appears to be not-implemented in "HCM" context (not ext_proc). Would that be a better solution? You'd use "raw_value" to the raw non-utf8 TLS value, which is then base64 encoded.

Reply:
@kyessenov I'm not sure if that would be a better solution, and I don't see any references to `HeaderOperations` in the code. It seems like it might be more work to implement this solution. Do you have any objections to encoding the TLVs at this level?

Reply:
I was referring to the top-level comment. It should be raw_value since the output is non-utf8. But then I noticed that raw_value was not implemented for header operations - IMHO it should, and we can create a patch for that.

```
      request_headers_to_add:
            - header:
                key: proxy-protocol-tlv
                raw_value: "%FILTER_STATE(envoy.network.proxy_protocol_options:FIELD:<tlv_type>)%"
```

Reply:
Got it, thanks for clarifying @kyessenov ! I understand the approach now. I wonder if this work should be handled in this patch, or in a separate issue? 

I don't have as much context on that part of envoy (this is my first attempt to commit upstream), so I'm not sure if I'm the best person to take this on. However, I'm happy to try if you are not willing to merge this current patch upstream without the proposed `raw_value` support.

Reply:
It should be a separate patch, I opened https://github.com/envoyproxy/envoy/issues/29590 to track it. This PR won't work because you have to own the underlying encoded string in the filter state. You can cache it and expose that way.